File: benchmarks/complex/mergesort/mergesort.rs
use std::time::Instant;

const N: usize = 2_000_000;
static mut RNG: i64 = 42;

fn lcg() -> i64 { unsafe { RNG = (RNG*1103515245+12345)&2147483647; RNG } }

fn merge_run(arr: &mut [i64], tmp: &mut [i64], lo: usize, mid: usize, hi: usize) {
    let (mut i, mut j, mut k) = (lo, mid, lo);
    while k < hi {
        if i < mid && (j >= hi || arr[i] <= arr[j]) { tmp[k] = arr[i]; i += 1; }
        else { tmp[k] = arr[j]; j += 1; }
        k += 1;
    }
    arr[lo..hi].copy_from_slice(&tmp[lo..hi]);
}

fn main() {
    let mut arr = vec![0i64; N];
    let mut tmp = vec![0i64; N];
    for x in arr.iter_mut() { *x = lcg(); }
    let t0 = Instant::now();
    let mut w = 1;
    while w < N {
        let mut lo = 0;
        while lo < N {
            let mid = (lo + w).min(N);
            let hi = (lo + 2*w).min(N);
            if mid < hi { merge_run(&mut arr, &mut tmp, lo, mid, hi); }
            lo += 2*w;
        }
        w *= 2;
    }
    let ms = t0.elapsed().as_millis();
    let ok = arr.windows(2).all(|w| w[0] <= w[1]);
    let chk: i64 = arr.iter().fold(0i64, |a,&x| (a+x)&0xffffffff);
    println!("n: {}\nsorted: {}\nfirst: {}\nlast: {}\nchecksum: {}\ntime: {} ms",
             N, ok as i32, arr[0], arr[N-1], chk, ms);
}

File: benchmarks/complex/mergesort/mergesort.go
package main

import (
	"fmt"
	"time"
)

const N = 2000000

var arr [N]int64
var tmp [N]int64
var rng int64 = 42

func lcg() int64 { rng = (rng*1103515245 + 12345) & 2147483647; return rng }

func mergeRun(lo, mid, hi int) {
	i, j, k := lo, mid, lo
	for k < hi {
		if i < mid && (j >= hi || arr[i] <= arr[j]) { tmp[k] = arr[i]; i++ } else { tmp[k] = arr[j]; j++ }
		k++
	}
	for x := lo; x < hi; x++ { arr[x] = tmp[x] }
}

func main() {
	for i := 0; i < N; i++ { arr[i] = lcg() }
	t0 := time.Now()
	for w := 1; w < N; w *= 2 {
		for lo := 0; lo < N; lo += 2 * w {
			mid := lo + w; if mid > N { mid = N }
			hi := lo + 2*w; if hi > N { hi = N }
			if mid < hi { mergeRun(lo, mid, hi) }
		}
	}
	ms := time.Since(t0).Milliseconds()
	ok := 1; for i := 1; i < N; i++ { if arr[i] < arr[i-1] { ok = 0; break } }
	var chk int64; for i := 0; i < N; i++ { chk = (chk + arr[i]) & 0xffffffff }
	fmt.Printf("n: %d\nsorted: %d\nfirst: %d\nlast: %d\nchecksum: %d\ntime: %d ms\n",
		N, ok, arr[0], arr[N-1], chk, ms)
}
